packaging/package-indexer/indexer/rpm_indexer.py

```diff
@@ -0,0 +1,304 @@
+        incoming_sub_dir: Path,
+        dist_dir: Path,
+        cdn: CDN,
+        gpg_key_path: Path,
+        gpg_key_passphrase: Optional[str],
+    ) -> None:
+        self.__gpg_key_path = gpg_key_path
+        self.__gpg_key_passphrase = gpg_key_passphrase
+        super().__init__(
+            incoming_remote_storage_synchronizer=incoming_remote_storage_synchronizer,
+            indexed_remote_storage_synchronizer=indexed_remote_storage_synchronizer,
+            incoming_sub_dir=incoming_sub_dir,
+            indexed_sub_dir=Path("rpm", dist_dir),
+            cdn=cdn,
+        )
+
+    def __move_files_from_incoming_to_indexed(self, incoming_dir: Path, indexed_dir: Path) -> None:
+        for file in filter(lambda path: path.is_file(), incoming_dir.rglob("*")):
+            relative_path = file.relative_to(incoming_dir)
+            platform = relative_path.parent
+            file_name = relative_path.name
+
+            new_path = Path(indexed_dir, platform, "x86_64", file_name)
+
+            self._log_info("Moving file.", src_path=str(file), dst_path=str(new_path))
+
+            new_path.parent.mkdir(parents=True, exist_ok=True)
+            utils.move_file_without_overwrite(file, new_path)
+
+    def _prepare_indexed_dir(self, incoming_dir: Path, indexed_dir: Path) -> None:
+        self.__move_files_from_incoming_to_indexed(incoming_dir, indexed_dir)
+
+    def __get_gpg_key_email(self, gnupghome_dir: Path) -> str:
+        with TemporaryFile(mode="w+", encoding="utf-8") as f:
+            utils.execute_command(
+                ["gpg", "--list-secret-keys", "--with-colons"],
+                stdout=f,
+                env={"GNUPGHOME": str(gnupghome_dir)},
+            )
+            f.seek(0)
+
+            for line in f.readlines():
+                if line.startswith("uid"):
+                    return line.split(":")[9]
+
+        raise Exception("Failed to get GPG key email.")
+
+    def __rpm_addsign(self, gnupghome_dir: Path, indexed_dir: Path) -> None:
+        email = self.__get_gpg_key_email(gnupghome_dir)
+
+        sign_cmd = r"%{__gpg} gpg --no-tty --no-options --batch --no-armor -u \"%{_gpg_name}\" -sbo %{__signature_filename} %{__plaintext_filename}'"
+
+        rpmmacros_path = Path(gnupghome_dir, ".rpmmacros").absolute()
+        rpmmacros_path.write_text(
+            "%_signature gpg\n"
+            f"%_gpg_name {email}\n"
+            f"%_gpg_path {str(gnupghome_dir)}\n"
+            f"%__gpg {shutil.which('gpg')}\n"
+            f"%__gpg_sign_cmd {sign_cmd}\n"
+        )
+
+        env = {"GNUPGHOME": str(gnupghome_dir), "GPG_TTY": ""}
+
+        for rpm_file in indexed_dir.rglob("*.rpm"):
+            command = ["rpm", "--macros", str(rpmmacros_path), "--addsign", str(rpm_file.absolute())]
+
+            self._log_info("Signing rpm.", rpm_file=str(rpm_file))
+            utils.execute_command(command, env=env)
+
+    def __create_repomd_xml_asc_file(self, gnupghome_dir: Path, repo_path: Path) -> None:
```

Comment:
The naming around RPM repos confuses me.
Is this yum-specific? Does the same format apply to `dnf` and `zypper`?

Reply:
As far as I can understand the only structure that is needed is to have a dir with `.rpm`s, and a `repodata` subdir, for the meta files (`createrepo` makes this for us).

On how you find the dir for your OS and arch:
The dir structure does not matter (e.g. `rpm/nightly/almalinux-8/x86_64/`), but it is useful to name them in a way, that the yum variables resolve to: https://docs.redhat.com/en/documentation/red_hat_enterprise_linux/6/html/deployment_guide/sec-using_yum_variables#sec-Using_Yum_Variables

I tested with `dnf`, and it uses the same `/etc/yum.repos.d` dir, and works. Did not test `zypper`, though.

So I guess parts of this is yum specific, but not sure how much.

```
# dnf install axosyslog                                                                                                                                                                         
AlmaLinux 9 - CRB                                                                                                                                                                    3.7 MB/s | 2.7 MB     00:00     
Extra Packages for Enterprise Linux 9 - x86_64                                                                                                                                        15 MB/s |  23 MB     00:01     
Dependencies resolved.                                                                                                                                                                                               
=====================================================================================================================================================================================================================
 Package                                             Architecture                           Version                                                                  Repository                                 Size 
=====================================================================================================================================================================================================================
Installing:
 axosyslog                                           x86_64                                 0.8.1.18143.g7f5f3f5-2+20241126T135514                                   axosyslog                                 1.1 M
Installing dependencies:
 ivykis                                              x86_64                                 0.42.4-7.el9                                                             epel                                       46 k
 libnet                                              x86_64                                 1.2-7.el9                                                                appstream                                  57 k
 logrotate                                           x86_64                                 3.18.0-8.el9                                                             baseos                                     74 k
Installing weak dependencies:
 axosyslog-logrotate                                 x86_64                                 0.8.1.18143.g7f5f3f5-2+20241126T135514                                   axosyslog                                 8.6 k

Transaction Summary
=====================================================================================================================================================================================================================
Install  5 Packages

Total download size: 1.3 M
...
```

Reply:
Thanks.

The man of createrepo says `repomd (xml-based rpm metadata)`, which also clears a few things for me.